### PR TITLE
Update files flows for Platforms

### PR DESCRIPTION
**Pull request by martyna-lachut-cko**
# Description of changes in PR

[//]: # 'Update file upload flow and added files retrieval documentation for Platforms'

## Checklist

- [X] Added a changelog entry to the `changelog.md` file in either `abc_spec` or `nas_spec`
- [X] Contacted the Tech Docs team about any corresponding guides that need to be updated for www.checkout.com/docs

**Contributing options**: Look at [our documentation](https://checkout.atlassian.net/wiki/spaces/PD/pages/2169506663/API+ref+publication+process) for options to contribute to the API reference.

### CODE REVIEW

File: nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml [hunk 1]
```diff
@@ -0,0 +1,50 @@
+type: object
+properties:
+  id:
+    type: string
+    description: The id of the file
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml:5]:
```suggestion
    description: The ID of the file
```

File: nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+    example: file_6lbss42ezvoufcb2beo76rvwly
+  status:
+    type: string
+    description: The current status of the file
+    example: verified
+  status_reasons:
+    type: array
+    description: Reasons for Invalid status. Null, if status is other than Invalid
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml:13]:
```suggestion
    description: If `status` is `Invalid`, returns the reasons the file was invalid. If `status` is any other value, returns `Null`.
```

File: nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+    example: ["InvalidMimeType"]
+  file_name:
+    type: string
+    description: The name of the file
+    example: file.jpeg
+  size:
+    type: number
+    description: The size of the document in KB
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml:21]:
```suggestion
    description: The size of the file, in KB
```

File: nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    example: 1024
+  mime_type:
+    type: string
+    description: The MIME type of the document
+    example: application/pdf
+  uploaded_on:
+    type: string
+    format: date-time
+    description: The ISO 8601 UTC datetime when the file was uploaded
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml:30]:
```suggestion
    description: The date and time the file was uploaded, in ISO 8601 UTC format
```

File: nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    example: "2020-12-01T15:01:01Z"
+  purpose:
+    type: string
+    description: The purpose of this document
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml:34]:
```suggestion
    description: The purpose of the file
```

Comment (sebastian-garcia-cardona-cko) [reply nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml:34]: Is this a value that the user would have supplied themselves in the request?

File: nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    type: string
+    description: The name of the file
+    example: file.jpeg
+  size:
+    type: number
+    description: The size of the document in KB
+    example: 1024
+  mime_type:
+    type: string
+    description: The MIME type of the document
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml:25]: What 
```suggestion
    description: The MIME type of the file
```

File: nas_spec/components/schemas/Platforms/PlatformsFileUpload.yaml [hunk 7]
```diff
@@ -13,7 +13,7 @@ properties:
       - financial_verification
     example: 'identity_verification'
   
-  path:
+  entity_id:
     type: string
-    description: The <strong>local</strong> path of the file to upload, and its type.
-    example: 'file=@/path/passport.png;type=image/png'
+    description: The id of the entity this request relates to
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileUpload.yaml:18]:
```suggestion
    description: The ID of the entity that the request relates to
```

File: nas_spec/components/schemas/Platforms/PlatformsFileUploadResponse.yaml [hunk 8]
```diff
@@ -5,6 +5,18 @@ properties:
     type: string
     description: The file identifier
     example: 'file_6lbss42ezvoufcb2beo76rvwly'
+  status:
+    type: string
+    description: The current status of the file
+    example: created
+  maximum_size_in_bytes:
+    type: number
+    description: The maxiumum allowed file size in bytes
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileUploadResponse.yaml:14]:
```suggestion
    description: The maximum file size allowed, in bytes
```

File: nas_spec/components/schemas/Platforms/PlatformsFileUploadResponse.yaml [hunk 9]
```diff
@@ -5,6 +5,18 @@ properties:
...
+    example: 4194304
+  document_types_for_purpose:
+    type: array
+    description: The allowed file mime types for the document purpose provided on the initial file upload request
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileUploadResponse.yaml:18]:
```suggestion
    description: The MIME file types allowed for the document purpose, provided on the initial file upload request
```

File: nas_spec/components/schemas/Platforms/PlatformsFileUploadResponse.yaml [hunk 10]
```diff
@@ -13,3 +25,7 @@ properties:
         properties:
           href:
             example: 'https://files.checkout.com/files/file_6lbss42ezvoufcb2beo76rvwly'
+      upload:
+        description: The file upload link to send the file to, in a data-binary type request
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileUploadResponse.yaml:29]: Would replacing the first half of the sentence with the below still be accurate?

> The URL to upload the file to

Also, could you clarify what you mean by 'in a data-binary type request' please?

File: nas_spec/paths/platforms-files@{file_id}.yaml [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  operationId: retrieveAFile        
+  description: >-
+    Retrieve information about a previously uploaded file. 
+  parameters:
+    - in: path
+      name: file_id
+      schema:
+        type: string
+      required: true
+      description: The file identifier. It is always prefixed by `file_`.
```
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/paths/platforms-files@{file_id}.yaml:20]:
```suggestion
      description: The ID of the file. The value is always prefixed by `file_`.
```

**Review by sebastian-garcia-cardona-cko — CHANGES_REQUESTED** [review]

File: nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml (hunk 5 above)
Comment (martyna-lachut-cko) [reply nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml:34]: Yes exactly, they will have provided this value on the initial upload request.

File: nas_spec/components/schemas/Platforms/PlatformsFileUploadResponse.yaml (hunk 10 above)
Comment (sebastian-garcia-cardona-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileUploadResponse.yaml:29]:
```suggestion
        description: The URL to upload the file to, in a data-binary type request
```

Comment (sebastian-garcia-cardona-cko) [reply nas_spec/components/schemas/Platforms/PlatformsFileUploadResponse.yaml:29]: I see, will keep that copy as-is then. Have suggested a change for the first part of the sentence

File: nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml (hunk 5 above)
Comment (sebastian-garcia-cardona-cko) [reply nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml]: Thanks, have updated the suggestion

**Review by sebastian-garcia-cardona-cko — APPROVED** [review]

**Comment by Armando-Ibarra** [comment]
@martyna-lachut-cko any particular reason why this endpoint 
https://www.checkout.com/docs/platforms/onboarding/api/full/upload-a-file#Attach_the_file

Isn't on the API reference? Since this is an important step, it's odd that we do not expose the `PUT` operation.

**Comment by martyna-lachut-cko** [comment]
> @martyna-lachut-cko any particular reason why this endpoint https://www.checkout.com/docs/platforms/onboarding/api/full/upload-a-file#Attach_the_file
> 
> Isn't on the API reference? Since this is an important step, it's odd that we do not expose the `PUT` operation.

Hi Armando, the PUT request doesn't pass through the Files API, it goes straight to S3, therefore I didn't include it in the API reference. What do you think @sebastian-garcia-cardona-cko  ?

File: nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml [hunk 12]
```diff
@@ -0,0 +1,50 @@
+type: object
+properties:
+  id:
+    type: string
+    description: The ID of the file.
+    example: file_6lbss42ezvoufcb2beo76rvwly
+  status:
+    type: string
+    description: The current status of the file.
+    example: verified
```
Comment (natasha-moore-cko) [inline nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml:10]: Hey @martyna-lachut-cko, I'm just seeing this PR now as I was off last week. Should we change the `status` value in this example from `verified` to `invalid` (due to `status_reasons` being `InvalidMimeType`)?

image

Comment (natasha-moore-cko) [reply nas_spec/components/schemas/Platforms/PlatformsFileRetrieveResponse.yaml:10]: A couple more questions:
Should we change the `file_name` value in this example to `file.pdf` to match the mime type?
Should we add `file_name` to response examples in the docs [here](https://www.checkout.com/docs/platforms/onboarding/api/full/upload-a-file#Response_examples)?

**Comment by Armando-Ibarra** [comment]
> > @martyna-lachut-cko any particular reason why this endpoint https://www.checkout.com/docs/platforms/onboarding/api/full/upload-a-file#Attach_the_file
> > Isn't on the API reference? Since this is an important step, it's odd that we do not expose the `PUT` operation.
> 
> Hi Armando, the PUT request doesn't pass through the Files API, it goes straight to S3, therefore I didn't include it in the API reference. What do you think @sebastian-garcia-cardona-cko ?

Thanks @martyna-lachut-cko, but seems that either way is passing through `files` API, because you have this PUT
`curl --location --request PUT “https://files.sandbox.checkout.com/files/ent_e5fxiiuec7yasswclx27zbpbcc/file_bbmcujt2fxmyh5pmeuz2bvvxd?AWSAccessKeyId=XX&Expires=1673456742&Signature=fg9Eq3mQBXrlnbeWCTyrURlBxQ%3X”
--data-binary '@/C:/Users/Test/test.pdf'`

And you need to call first `files.sandbox.checkout.com` even if internally this redirects to AWS

Also, this is not code friendly, consider that most of the merchants uses SDK or his own implementation for calling API's, which means that a http call must be created in order to accomplish this, this increases the complexity in code just to upload a single file.

**Comment by martyna-lachut-cko** [comment]
> > > @martyna-lachut-cko any particular reason why this endpoint https://www.checkout.com/docs/platforms/onboarding/api/full/upload-a-file#Attach_the_file
> > > Isn't on the API reference? Since this is an important step, it's odd that we do not expose the `PUT` operation.
> > 
> > 
> > Hi Armando, the PUT request doesn't pass through the Files API, it goes straight to S3, therefore I didn't include it in the API reference. What do you think @sebastian-garcia-cardona-cko ?
> 
> Thanks @martyna-lachut-cko, but seems that either way is passing through `files` API, because you have this PUT `curl --location --request PUT “https://files.sandbox.checkout.com/files/ent_e5fxiiuec7yasswclx27zbpbcc/file_bbmcujt2fxmyh5pmeuz2bvvxd?AWSAccessKeyId=XX&Expires=1673456742&Signature=fg9Eq3mQBXrlnbeWCTyrURlBxQ%3X” --data-binary '@/C:/Users/Test/test.pdf'`
> 
> And you need to call first `files.sandbox.checkout.com` even if internally this redirects to AWS
> 
> Also, this is not code friendly, consider that most of the merchants uses SDK or his own implementation for calling API's, which means that a http call must be created in order to accomplish this, this increases the complexity in code just to upload a single file.

Apologies, the upload link in the public docs and here is incorrect (it's being corrected), it should be: `https://s3.eu-west-1.amazonaws.com/mp-files-api-staging-sbox/ent_e5fxiiuec7yasswclx27zbpbcc/file_bbmcujt2fxmyh5pmeuz2bvvxd?AWSAccessKeyId=XX&Expires=1673456742&Signature=fg9Eq3mQBXrlnbeWCTyrURlBxQ%3X`.
Hopefully now it is clearer to see that the request goes directly to S3.

**Pull request merged**